crates/bevy_ecs/src/resource.rs

```diff
@@ -73,3 +79,89 @@ pub use bevy_ecs_macros::Resource;
     note = "consider annotating `{Self}` with `#[derive(Resource)]`"
 )]
 pub trait Resource: Send + Sync + 'static {}
+
+/// A marker component for the entity that stores the resource of type `T`.
+///
+/// This component is automatically inserted when a resource of type `T` is inserted into the world,
+/// and can be used to find the entity that stores a particular resource.
+///
+/// By contrast, the [`IsResource`] component is used to find all entities that store resources,
+/// regardless of the type of resource they store.
+///
+/// This component comes with a hook that ensures that at most one entity has this component for any given `R`:
+/// adding this component to an entity (or spawning an entity with this component) will despawn any other entity with this component.
+#[derive(Component, Debug)]
+#[require(Internal, IsResource)]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Component, Default))]
+pub struct ResourceEntity<R: Resource>(#[reflect(ignore)] PhantomData<R>);
+
+impl<R: Resource> Default for ResourceEntity<R> {
+    fn default() -> Self {
+        ResourceEntity(PhantomData)
+    }
+}
+
+/// A marker component for entities which store resources.
+///
+/// By contrast, the [`ResourceEntity<R>`] component is used to find the entity that stores a particular resource.
+/// This component is required by the [`ResourceEntity<R>`] component, and will automatically be added.
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(Reflect),
+    reflect(Component, Default, Debug)
+)]
+#[derive(Component, Default, Debug)]
+pub struct IsResource;
+
+/// Used in conjunction with [`ResourceEntity<R>`], when no type information is available.
+/// This is used by [`World::insert_resource_by_id`](crate::world::World).
+#[derive(Resource)]
+pub(crate) struct TypeErasedResource;
```

Comment:
This needs to be pub + Debug + Reflect: it's important that this show up in inspectors properly.